IAM module returns created keys

mdawid92 opened the pull request:
##### ISSUE TYPE
 - Bugfix Pull Request


##### COMPONENT NAME
cloud/amazon/iam

##### ANSIBLE VERSION
```
ansible 2.3.0 (iam-keys 3aee11e48b) last updated 2017/02/10 08:26:38 (GMT +000)
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
```

##### SUMMARY

Migrating  https://github.com/ansible/ansible-modules-core/pull/3385 by @defionscode

Fixes issues in https://github.com/ansible/ansible-modules-core/issues/3346

Fixes issue https://github.com/ansible/ansible-modules-core/issues/2821 where newly created keys were not appearing in the returned output by the module. 

Adds list with created keys to output:
```
{"changed": true, "created_keys": [{"access_key_id": "FAKE3FZHKFAKE2QA", "access_key_selector": "HMAC", "create_date": "2016-04-07T20:09:11.799Z", "secret_access_key": "fake_key_alsjdflkasdfksf;lksjdfkls", "status": "Active", "user_name": "pullrequest_testing"}], "groups": null, "keys": {"FAKE3FZHKFAKE2QA": "Active"}, "user_name": "pullrequest_testing"}```
```

mdawid92 committed:
return new key, porting https://github.com/ansible/ansible-modules-core/pull/3385/ by defionscode

mdawid92 committed:
fix python3 compatibility

mdawid92 commented:
ready_for_review

fivethreeo commented:
Unindent line 390 to 397 and this is perfect 👍

mdawid92 committed:
fixed indentation

mdawid92 commented:
Thanks, fixed.

djui commented:
What is the status of this PR? Once it's merged, which version of Ansible will be required?

Is there no other way to retrieve the access key secret

felipe1982 commented:
Would be really nice to have this merged.

Ref: #2821 **iam module not very useful for managing access keys**

s-hertel commented:
Reviewing this today.

@felipe1982 Is that you issue you meant to link?

s-hertel commented:
@mdawid92 I notice that if I make two keys one after another, when I make a second one I do not get the same kind of return value for the new key. The first time it comes in `user_meta.access_keys` and the second time in `created_keys`. Can we make that uniform so the return structure isn't a surprise? Also...at some point return values for this module ought to be documented...but that looks like it will be a mess and can probably be a follow up PR to this fix.

mdawid92 commented:
@s-hertel Thanks for review. Yes, it should be uniform. The one with `user_meta.access_keys` is returned while creating new user (it's already in ansible), second one while updating user (new in this PR, but there's no `user_meta` while updating). Which version would you prefer?
I also noticed that `keys` field in this two cases are different - while creating it is more verbose.

felipe1982 commented:
@mdawid92 I am quite new to Ansible, however may I suggest *"principle of least astonishment"*? So for the problem of whether to return `user_meta` or `created_keys` should depend on what a proficient Ansible user might expect ....

I was also required to write this "funny" code to get a NEW access key id for an EXISTING user:
```
ok: [localhost] => {
    "changed": false,
    "created_keys": [],
    "groups": null,
    "invocation": {
        "module_args": {
            "access_key_ids": null,
            "access_key_state": null,
            "aws_access_key": null,
            "aws_secret_key": null,
            "ec2_url": null,
            "groups": null,
            "iam_type": "user",
            "key_count": 1,
            "name": "StagingAuth0InstanceRole",
            "new_name": null,
            "new_path": null,
            "password": null,
            "path": "/",
            "profile": null,
            "region": null,
            "security_token": null,
            "state": "present",
            "trust_policy": null,
            "trust_policy_filepath": null,
            "update_password": "always",
            "validate_certs": true
        }
    },
    "keys": {
        "AKIAJJ7UGZEZLPJUU3OQ": "Active"
    },
    "user_name": "StagingAuth0InstanceRole"
```
```
- name: "Get user's data (or make a user if not present)"
  iam:
    iam_type: user
    name: 'myname'
    state: "present"
  register: iam_user

- name: Extract current access key id
  set_fact:
    iam_user_access_key_id: "{{ iam_user.get('keys').keys() }}"
  when: iam_user|succeeded
```
An easier (that is to say a Jinja-native or Ansible-native way) to extract the access_key_id would be great, instead of resorting to Python dict methods.

s-hertel commented:
@mdawid92 Sorry for not getting back to this sooner. Maybe the best course of action is to add user_meta.access_keys to the returned data when you're getting existing keys, so it's accessible both when you've created a key and when getting preexisting keys (but have the created_keys return value too when you've created them).

mdawid92 committed:
added user_meta field

mdawid92 commented:
Sorry for long time to update pull request. I've added user_meta.access_keys to updating users/creating new keys. If there is existing key, it is also added to that field:
```changed: [localhost] => {
    "changed": true,
    "created_keys": [
        {
            "access_key_id": "AKIAJNX3JJAMYB4KRG4A",
            "access_key_selector": "HMAC",
            "create_date": "2017-06-20T09:16:08.036Z",
            "secret_access_key": "c*****x",
            "status": "Active",
            "user_name": "test_5"
        }
    ],
    "groups": null,
    "keys": {
        "AKIAIXAIWLTL2M6KFZ5Q": "Active",
        "AKIAJNX3JJAMYB4KRG4A": "Active"
    },
    "user_meta": {
        "access_keys": [
            {
                "access_key_id": "AKIAJNX3JJAMYB4KRG4A",
                "access_key_selector": "HMAC",
                "create_date": "2017-06-20T09:16:08.036Z",
                "secret_access_key": "c*****x",
                "status": "Active",
                "user_name": "test_5"
            },
            {
                "access_key_id": "AKIAIXAIWLTL2M6KFZ5Q",
                "status": "Active"
            }
        ]
    },
    "user_name": "test_5"
}

```
We could also get "create_date", "user_name" for preexisting keys with small tweaking.

s-hertel commented:
shipit

ryansb commented:
Rerunning CI then this should be good to merge.
